xdsl/transforms/mpi_to_llvm.py

```diff
@@ -0,0 +1,525 @@
+        """
+        This method lowers mpi.recv operations
+
+        MPI_Recv signature:
+
+        int MPI_Recv(void *buf, int count, MPI_Datatype datatype, int source, int tag,
+             MPI_Comm comm, MPI_Status *status)
+        """
+        count_ops, count_ssa_val = self._emit_memref_counts(op.buffer)
+
+        ops, new_results, status = self._emit_mpi_status_obj(
+            len(op.results) == 0)
+
+        # TODO: I really hate this dance just to make pyright happy
+        #       imo this makes code *less* readable.
+        #       The _MemRefTypeElement is bound to Attribute, so
+        #       op.buffer.typ.element_type is ALLWAYS at least Attribute!
+        assert isinstance(op.buffer.typ, memref.MemRefType)
+        memref_elm_typ = cast(memref.MemRefType[Attribute],
+                              op.buffer.typ).element_type
+
+        return [
+            *count_ops,
+            *ops,
+            *(ptr := self._memref_get_llvm_ptr(op.buffer))[0],
+            datatype := self._emit_mpi_type_load(memref_elm_typ),
+            tag := arith.Constant.from_int_and_width(op.tag.value.data,
+                                                     mpi.t_int),
+            comm_global :=
+            arith.Constant.from_int_and_width(self.info.mpi_comm_world_val,
+                                              mpi.t_int),
+            func.Call.get(self._mpi_name(op), [
+                ptr[1], count_ssa_val, datatype, op.source, tag, comm_global,
+                status
+            ], [mpi.t_int]),
+        ], new_results
+
+    # Miscellaneous
+
+
+class MpiAddExternalFuncDefs(RewritePattern):
+    mpi_func_call_names = set(_MPIToLLVMRewriteBase.MPI_SYMBOL_NAMES.values())
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, module: builtin.ModuleOp,
+                          rewriter: PatternRewriter, /):
+        # collect all func calls to MPI functions
+        funcs_to_emit: dict[str, tuple[list[Attribute],
+                                       list[Attribute]]] = dict()
+
+        @op_type_rewrite_pattern
+        def match_func(op: func.Call, rewriter: PatternRewriter, /):
+            if op.callee.string_value() not in self.mpi_func_call_names:
+                return
+            funcs_to_emit[op.callee.string_value()] = (list(
+                arg.typ
+                for arg in op.arguments), list(res.typ for res in op.results))
+
+        PatternRewriteWalker(AnonymousRewritePattern(match_func))
+
+        # for each func found, add a FuncOp to the top of the module.
+        for name, types in funcs_to_emit.items():
+            arg, res = types
+            rewriter.insert_op_at_pos(func.FuncOp.external(name, arg, res),
+                                      module.body.blocks[0], 0)
+
+
+def mpi_to_llvm_lowering(ctx: MLContext, module: builtin.ModuleOp):
+    # TODO: how to get the lib info in here?
+    lib_info = MpiLibraryInfo()
+
+    # lower to func.call
+    walker1 = PatternRewriteWalker(
```

Comment:
(Just checking) did you give this a thought of how it exactly walks? (inside-out, reverse, recursive application...)

Reply:
I think I corrected that!